modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaAwareLeaseTracker.java

```diff
@@ -0,0 +1,131 @@
+
+    private final PlacementDriver delegate;
+    private final ReplicaService replicaService;
+
+    /** Resolver that resolves a node consistent ID to cluster node. */
+    private final ClusterNodeResolver clusterNodeResolver;
+
+
+    /**
+     * Constructor.
+     *
+     * @param delegate Delegate Placement Driver.
+     * @param replicaService Replica Service.
+     * @param clusterNodeResolver Cluster node resolver.
+     */
+    public ReplicaAwareLeaseTracker(PlacementDriver delegate, ReplicaService replicaService, ClusterNodeResolver clusterNodeResolver) {
+        this.delegate = delegate;
+        this.replicaService = replicaService;
+        this.clusterNodeResolver = clusterNodeResolver;
+    }
+
+    @Override
+    public void listen(PrimaryReplicaEvent evt, EventListener<? extends PrimaryReplicaEventParameters> listener) {
+        delegate.listen(evt, listener);
+    }
+
+    @Override
+    public void removeListener(PrimaryReplicaEvent evt, EventListener<? extends PrimaryReplicaEventParameters> listener) {
+        delegate.removeListener(evt, listener);
+    }
+
+    @Override
+    public CompletableFuture<ReplicaMeta> awaitPrimaryReplica(ReplicationGroupId groupId, HybridTimestamp timestamp, long timeout,
+            TimeUnit unit) {
+        return delegate.awaitPrimaryReplica(groupId, timestamp, timeout, unit);
+    }
+
+    @Override
+    public CompletableFuture<ReplicaMeta> awaitPrimaryReplicaForTable(
+            ReplicationGroupId groupId,
+            HybridTimestamp timestamp,
+            long timeout,
+            TimeUnit unit
+    ) {
+        ZonePartitionId zonePartitionId = (ZonePartitionId) groupId;
+
+        TablePartitionId tablePartitionId = new TablePartitionId(zonePartitionId.tableId(), zonePartitionId.partitionId());
+
+        return delegate.awaitPrimaryReplica(tablePartitionId, timestamp, timeout, unit)
+                .thenCompose(replicaMeta -> {
+                    ClusterNode leaseholderNode = clusterNodeResolver.getById(replicaMeta.getLeaseholderId());
+
+                    if (replicaMeta.subgroups().contains(tablePartitionId)) {
+                        return completedFuture(replicaMeta);
+                    }
+
+                    WaitReplicaStateMessage awaitReplicaReq = REPLICA_MESSAGES_FACTORY.waitReplicaStateMessage()
+                            .groupId(tablePartitionId)
+                            .enlistmentConsistencyToken(replicaMeta.getStartTime().longValue())
+                            .timeout(timeout)
+                            .build();
+
+                    return replicaService.invoke(leaseholderNode, awaitReplicaReq).thenApply((ignored) -> replicaMeta);
+                });
+    }
+
+    @Override
+    public CompletableFuture<ReplicaMeta> getPrimaryReplica(ReplicationGroupId replicationGroupId, HybridTimestamp timestamp) {
+        return delegate.getPrimaryReplica(replicationGroupId, timestamp);
```

Comment:
Why it's not implemented int the similar manner like awaitPrimaryReplica?

Reply:
It will be implemented in the task for the corresponding module where `getPrimaryReplica` is used